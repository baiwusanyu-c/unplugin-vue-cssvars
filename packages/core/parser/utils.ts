// Enums:
/** This is derived from `@vue/compiler-core` */
import type { BindingTypes as VueBindingTypes } from '@vue/compiler-dom'
export const CSSVarsBindingTypes = {
  /**
   * returned from data()
   */
  DATA: 'data' as VueBindingTypes.DATA,
  /**
   * declared as a prop ✅
   */
  PROPS: 'props' as VueBindingTypes.PROPS,
  /**
   * a local alias of a `<script setup>` destructured prop.
   * the original is stored in __propsAliases of the bindingMetadata object.
   */
  PROPS_ALIASED: 'props-aliased' as VueBindingTypes.PROPS_ALIASED,
  /**
   * a let binding (may or may not be a ref) ✅
   */
  SETUP_LET: 'setup-let' as VueBindingTypes.SETUP_LET,
  /**
   * a const binding that can never be a ref.
   * these bindings don't need `unref()` calls when processed in inlined
   * template expressions. ✅
   */
  SETUP_CONST: 'setup-const' as VueBindingTypes.SETUP_CONST,
  /**
   * a const binding that does not need `unref()`, but may be mutated. ✅
   */
  SETUP_REACTIVE_CONST: 'setup-reactive-const' as VueBindingTypes.SETUP_REACTIVE_CONST,
  /**
   * a const binding that may be a ref. ✅
   */
  SETUP_MAYBE_REF: 'setup-maybe-ref' as VueBindingTypes.SETUP_MAYBE_REF,
  /**
   * bindings that are guaranteed to be refs ✅
   */
  SETUP_REF: 'setup-ref' as VueBindingTypes.SETUP_REF,
  /**
   * declared by other options, e.g. computed, inject
   */
  OPTIONS: 'options' as VueBindingTypes.OPTIONS,
  /**
   * a literal constant, e.g. 'foo', 1, true ✅
   */
  LITERAL_CONST: 'literal-const' as VueBindingTypes.LITERAL_CONST,
} as const
