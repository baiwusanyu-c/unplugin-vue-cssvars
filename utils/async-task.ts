export function runAsyncTaskList(
  taskNum: number,
  taskFunc: (index: number) => Promise<Record<any, any> | void>) {
  const taskList = [] as Array<Promise<any>>
  for (let i = 0; i < taskNum; i++) {
    taskList.push(new Promise((resolve) => {
      resolve(taskFunc(i))
    }))
  }
  return taskList
}
