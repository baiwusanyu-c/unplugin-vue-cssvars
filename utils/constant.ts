export const NAME = 'unplugin-vue-cssvars'
export const SUPPORT_FILE_LIST = ['**/**.css']
export const SUPPORT_FILE_REG = /\.(css|sass|scss|styl|less)$/i
export const FG_IGNORE_LIST = ['**/node_modules/**', '**/dist/**', '**/.git/**']
export const SUPPORT_FILE = {
  CSS: 'css',
  LESS: 'less',
  SASS: 'sass',
  SCSS: 'scss',
  STYLUS: 'stylus',
  STYL: 'styl',
}
export const DEFAULT_INCLUDE_REG = [/\.vue$/]
export const DEFAULT_EXCLUDE_REG = [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/, /[\\/]\.dist[\\/]/]
export const JSX_TSX_REG = /^(.*\.(jsx|tsx)$)/
