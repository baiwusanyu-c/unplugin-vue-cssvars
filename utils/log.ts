import chalk from 'chalk'

export declare type TLog = 'error' | 'warning' | 'info' | 'success'

export const logType = {
  info: (msg: string, prefix = '') => chalk.blueBright.bold(`${prefix}${msg}`),
  error: (msg: string, prefix = '') => chalk.redBright.bold(`${prefix}${msg}`),
  warning: (msg: string, prefix = '') => chalk.yellowBright.bold(`${prefix}${msg}`),
  success: (msg: string, prefix = '') => chalk.greenBright.bold(`${prefix}${msg}`),
}

export const log = (type: TLog, msg: string, prefix = '[unplugin-vue-cssvars]:') => {
  console.log(logType[type](msg, prefix))
}
