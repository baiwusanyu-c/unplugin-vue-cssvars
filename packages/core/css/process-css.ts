import path from 'path'
import * as csstree from 'css-tree'
import { completeSuffix, transformSymbol } from '@unplugin-vue-cssvars/utils'
import { walkCSSTree } from './pre-process-css'
import type { ICSSFile, ICSSFileMap } from '../types'
import type { SFCDescriptor } from '@vue/compiler-sfc'

export const getCSSFileRecursion = (key: string, cssFiles: ICSSFileMap, cb: (res: ICSSFile) => void, matchedMark = new Set<string>()) => {
  // 避免循环引用
  if (matchedMark.has(key)) return
  const cssFile = cssFiles.get(key)
  if (cssFile) {
    matchedMark.add(key)
    cb(cssFile)
    if (cssFile.importer.size > 0) {
      cssFile.importer.forEach((value) => {
        getCSSFileRecursion(value, cssFiles, cb, matchedMark)
      })
    }
  }
}

/**
 * 遍历 sfc 的 style 标签内容
 * 根据其 ast，获取 @import 信息
 * @param descriptor
 * @param id transform's id
 * @param cssFiles
 */
export const createCSSModule = (descriptor: SFCDescriptor, id: string, cssFiles: ICSSFileMap) => {
  const importModule: Array<ICSSFile> = []
  // 遍历 sfc 的 style 标签内容
  for (let i = 0; i < descriptor.styles.length; i++) {
    const content = descriptor.styles[i].content
    const cssAst = csstree.parse(content)
    // 根据其 ast，获取 @import 信息
    walkCSSTree(cssAst, (importer) => {
      // 添加后缀
      // sfc中规则：如果@import 指定了后缀，则根据后缀，否则根据当前 script 标签的 lang 属性（默认css）
      let key = completeSuffix(transformSymbol(path.resolve(path.parse(id).dir, importer)), descriptor.styles[i].lang)
      // 如果 .scss 的 import 不存在，则用 css 的
      if (!cssFiles.get(key))
        key = completeSuffix(transformSymbol(path.resolve(path.parse(id).dir, importer)))

      // 根据 @import 信息，从 cssFiles 中，递归的获取所有在预处理时生成的 cssvars 样式
      getCSSFileRecursion(key, cssFiles, (res: ICSSFile) => {
        importModule.push(res)
      })
    }, { i: true, v: false })
  }
  return importModule
}
