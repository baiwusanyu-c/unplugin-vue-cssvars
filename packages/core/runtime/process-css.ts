import { parse, resolve } from 'path'
import { SUPPORT_FILE, completeSuffix, setTArray } from '@unplugin-vue-cssvars/utils'
import { parseImports } from '../parser'
import type { ICSSFile, ICSSFileMap } from '../types'
import type { SFCDescriptor } from '@vue/compiler-sfc'

export const getCSSFileRecursion = (
  key: string,
  cssFiles: ICSSFileMap,
  cb: (res: ICSSFile) => void,
  matchedMark = new Set<string>()) => {
  // 避免循环引用
  if (matchedMark.has(key)) return
  const cssFile = cssFiles.get(key)
  if (cssFile) {
    matchedMark.add(key)
    cb(cssFile)
    if (cssFile.importer.size > 0) {
      cssFile.importer.forEach((value) => {
        getCSSFileRecursion(value, cssFiles, cb, matchedMark)
      })
    }
  }
}

/**
 * 遍历 sfc 的 style 标签内容
 * 根据其 ast，获取 @import 信息
 * @param descriptor
 * @param id transform's id
 * @param cssFiles
 * @param server
 */
// TODO: unit test
export type TInjectCSSContent = Set<{ content: string, lang: string, styleTagIndex: number }>
export const getVBindVariableListByPath = (
  descriptor: SFCDescriptor,
  id: string,
  cssFiles: ICSSFileMap,
  server: boolean) => {
  const vbindVariable: Set<string> = new Set()
  const injectCSSContent: TInjectCSSContent = new Set()
  // 遍历 sfc 的 style 标签内容
  for (let i = 0; i < descriptor.styles.length; i++) {
    const content = descriptor.styles[i].content
    const lang = descriptor.styles[i].lang === SUPPORT_FILE.STYLUS ? SUPPORT_FILE.STYL : descriptor.styles[i].lang
    const idDirParse = parse(id)
    const parseImporterRes = parseImports(content)
    parseImporterRes.imports.forEach((res) => {
      const importerPath = resolve(idDirParse.dir, res.path)
      // 添加后缀
      // sfc中规则：如果@import 指定了后缀，则根据后缀，否则根据当前 script 标签的 lang 属性（默认css）
      let key = completeSuffix(importerPath, lang)
      // 如果 .scss 的 import 不存在，则用 css 的
      if (!cssFiles.get(key))
        key = completeSuffix(importerPath)

      // 根据 @import 信息，从 cssFiles 中，递归的获取所有在预处理时生成的 cssvars 样式
      getCSSFileRecursion(key, cssFiles, (res: ICSSFile) => {
        if (res.vBindCode) {
          !server && injectCSSContent.add({ content: res.content, lang: res.lang, styleTagIndex: i })
          res.vBindCode.forEach((vb) => {
            vbindVariable.add(vb)
          })
        }
      })
    })
  }
  return {
    vbindVariableListByPath: setTArray(vbindVariable),
    injectCSSContent,
  }
}
