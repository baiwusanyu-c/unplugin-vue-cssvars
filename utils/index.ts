import { SUPPORT_FILE, SUPPORT_FILE_REG } from './constant'

export * from './constant'

export const extend = <
  T extends Record<string, any>,
  U extends Record<string, any>>(
    objFir: T,
    objSec: U): T & U => {
  return Object.assign({}, objFir, objSec)
}

export const setTArray = <T>(set: Set<T>): Array<T> => { return [...set] }

export const isEmptyObj = (val: unknown) => JSON.stringify(val) === '{}'

export const transformSymbol = (path: string) => path.replaceAll('\\', '/')

export const completeSuffix = (fileName: string, suffix = SUPPORT_FILE.CSS) => {
  const transformSymbolRes = transformSymbol(fileName)
  return !(SUPPORT_FILE_REG.test(transformSymbolRes)) ? `${transformSymbolRes}.${suffix}` : transformSymbolRes
}
