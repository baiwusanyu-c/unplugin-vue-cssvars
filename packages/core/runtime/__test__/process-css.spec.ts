import { resolve } from 'path'
import { describe, expect, test, vi } from 'vitest'
import { transformSymbol } from '@unplugin-vue-cssvars/utils'
import { getCSSFileRecursion, getVBindVariableListByPath } from '../process-css'
import type { ICSSFile } from '../../types'
// TODO update
describe('process css', () => {
  test('getCSSFileRecursion: basic', () => {
    const mockEvt = vi.fn()
    const mockRes: Array<ICSSFile> = []
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(),
      vBindCode: {
        foo: new Set(['v-bind(foo)']),
      },
    })
    getCSSFileRecursion('css', 'foo', mockCssFiles, (v) => {
      mockEvt()
      mockRes.push(v)
    })
    expect(mockRes.length).toBe(1)
    expect(mockEvt).toBeCalledTimes(1)
    expect(mockRes[0].vBindCode).toMatchObject({
      foo: new Set(['v-bind(foo)']),
    })
  })

  test('getCSSFileRecursion: unmatched key', () => {
    const mockEvt = vi.fn()
    const mockRes = []
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(['foo1', 'foo2']),
    })
    getCSSFileRecursion('css', 'bar', mockCssFiles, (v) => {
      mockEvt()
      mockRes.push(v)
    })
    expect(mockRes.length).toBe(0)
    expect(mockEvt).not.toBeCalled()
  })

  test('getCSSFileRecursion: recursion', () => {
    const mockEvt = vi.fn()
    const mockRes: Array<ICSSFile> = []
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(['bar']),
      vBindCode: {
        foo: new Set(['v-bind(foo)']),
      },
    })
    mockCssFiles.set('bar', {
      importer: new Set(),
      vBindCode: {
        bar: new Set(['v-bind(bar)']),
      },
    })
    getCSSFileRecursion('css', 'foo', mockCssFiles, (v) => {
      mockEvt()
      mockRes.push(v)
    })
    expect(mockRes.length).toBe(2)
    expect(mockEvt).toBeCalledTimes(2)
    expect(mockRes[0].vBindCode).toMatchObject({
      foo: new Set(['v-bind(foo)']),
    })
    expect(mockRes[1].vBindCode).toMatchObject({
      bar: new Set(['v-bind(bar)']),
    })
  })

  test('getCSSFileRecursion: circular dependencies', () => {
    const mockEvt = vi.fn()
    const mockRes: Array<ICSSFile> = []
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(['bar']),
      vBindCode: {
        foo: new Set(['v-bind(foo)']),
      },
    })
    mockCssFiles.set('bar', {
      importer: new Set(['foo']),
      vBindCode: {
        bar: new Set(['v-bind(bar)']),
      },
    })
    getCSSFileRecursion('css', 'foo', mockCssFiles, (v) => {
      mockEvt()
      mockRes.push(v)
    })
    expect(mockRes.length).toBe(2)
    expect(mockEvt).toBeCalledTimes(2)
    expect(mockRes[0].vBindCode).toMatchObject({
      foo: new Set(['v-bind(foo)']),
    })
    expect(mockRes[1].vBindCode).toMatchObject({
      bar: new Set(['v-bind(bar)']),
    })
  })

  test('getVBindVariableListByPath: basic', () => {
    const mockCssFiles = new Map()
    const mockCSSFilesContent = {
      importer: new Set(),
      vBindCode: ['fooColor'],
      content: 'content foo color',
      lang: 'scss',
    }
    mockCssFiles.set(transformSymbol(resolve('/play/src/assets/test.css')), mockCSSFilesContent)
    const mockDescriptor = {
      styles: [{
        content: '@import "./assets/test";\n'
          + ' div {\n'
          + '   color: v-bind(color2);\n'
          + ' }',
      }],
    }
    const mockId = transformSymbol(resolve('/play/src/App.vue'))
    const res = getVBindVariableListByPath(mockDescriptor as any, mockId, mockCssFiles, false)
    expect(res.vbindVariableListByPath).toMatchObject(['fooColor'])
    expect([...res.injectCSSContent]).toMatchObject([{
      content: 'content foo color',
      lang: 'scss',
    }])
    expect(res).matchSnapshot()
  })

  test('getVBindVariableListByPath: server is true', () => {
    const mockCssFiles = new Map()
    const mockCSSFilesContent = {
      importer: new Set(),
      vBindCode: ['fooColor'],
      content: 'content foo color',
      lang: 'scss',
    }
    mockCssFiles.set(transformSymbol(resolve('/play/src/assets/test.css')), mockCSSFilesContent)
    const mockDescriptor = {
      styles: [{
        content: '@import "./assets/test";\n'
          + ' div {\n'
          + '   color: v-bind(color2);\n'
          + ' }',
      }],
    }
    const mockId = transformSymbol(resolve('/play/src/App.vue'))
    const res = getVBindVariableListByPath(mockDescriptor as any, mockId, mockCssFiles, true)
    expect(res.vbindVariableListByPath).toMatchObject(['fooColor'])
    expect([...res.injectCSSContent]).toMatchObject([])
    expect(res).matchSnapshot()
  })

  test('createCSSModule: no file with lang', () => {
    const mockCssFiles = new Map()
    const mockCSSFilesContent = {
      importer: new Set(),
      vBindCode: ['fooColor'],
    }
    mockCssFiles.set(transformSymbol(resolve('/play/src/assets/test.css')), mockCSSFilesContent)
    const mockDescriptor = {
      styles: [{
        lang: 'scss',
        content: '@import "./assets/test";\n'
          + ' div {\n'
          + '   color: v-bind(color2);\n'
          + ' }',
      }],
    }
    const mockId = transformSymbol(resolve('/play/src/App.vue'))
    const res = getVBindVariableListByPath(mockDescriptor as any, mockId, mockCssFiles, true)
    expect(res.vbindVariableListByPath).toMatchObject(['fooColor'])
    expect(res).matchSnapshot()
  })

  test('createCSSModule: multiple style tag', () => {
    const mockCssFiles = new Map()
    const mockCSSFilesContent = {
      importer: new Set(),
      vBindCode: ['fooColor'],
    }
    mockCssFiles.set(transformSymbol(resolve('/play/src/assets/test.css')), mockCSSFilesContent)
    const mockCSSFilesContent2 = {
      importer: new Set(),
      vBindCode: ['barColor'],
    }
    mockCssFiles.set(transformSymbol(resolve('/play/src/assets/test2.css')), mockCSSFilesContent2)
    const mockDescriptor = {
      styles: [{
        content: '@import "./assets/test";\n'
          + ' div {\n'
          + '   color: v-bind(color);\n'
          + ' }',
      },
      {
        content: '@import "./assets/test2";\n'
          + ' div {\n'
          + '   color: v-bind(color2);\n'
          + ' }',
      }],
    }
    const mockId = resolve('/play/src/App.vue')
    const res = getVBindVariableListByPath(mockDescriptor as any, mockId, mockCssFiles, true)
    expect(res.vbindVariableListByPath).toMatchObject(['fooColor', 'barColor'])
    expect(res).matchSnapshot()
  })

  test('createCSSModule: no style tag', () => {
    const mockCssFiles = new Map()
    mockCssFiles.set('foo', {
      importer: new Set(),
      vBindCode: ['fooColor'],
    })
    const mockDescriptor = {
      styles: [],
    }
    const res = getVBindVariableListByPath(mockDescriptor as any, 'foo', mockCssFiles, true)
    expect(res.vbindVariableListByPath.length).toBe(0)
  })
})
