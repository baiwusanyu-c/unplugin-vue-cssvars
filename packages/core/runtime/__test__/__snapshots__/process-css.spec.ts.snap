// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`process css > createCSSModule: multiple style tag 1`] = `
{
  "injectCSSContent": Set {},
  "vbindVariableListByPath": [
    "fooColor",
    "barColor",
  ],
}
`;

exports[`process css > createCSSModule: no file with lang 1`] = `
{
  "injectCSSContent": Set {},
  "vbindVariableListByPath": [
    "fooColor",
  ],
}
`;

exports[`process css > getVBindVariableListByPath: basic 1`] = `
{
  "injectCSSContent": Set {
    {
      "content": "content foo color",
      "lang": "scss",
    },
  },
  "vbindVariableListByPath": [
    "fooColor",
  ],
}
`;

exports[`process css > getVBindVariableListByPath: server is true 1`] = `
{
  "injectCSSContent": Set {},
  "vbindVariableListByPath": [
    "fooColor",
  ],
}
`;
